src/stores/AppController2.ts

```diff
@@ -0,0 +1,215 @@
+import WatchHistoryController from './WatchHistoryController';
+import FavoritesController from './FavoritesController';
+
+import type { Config } from '#types/Config';
+import { IntegrationInfo, useConfigStore } from '#src/stores/ConfigStore';
+import ConfigService from '#src/services/config.service';
+import { Settings, useSettingsStore } from '#src/stores/SettingsStore';
+import { logDev } from '#src/utils/common';
+import { OTT_GLOBAL_PLAYER_ID, CONFIG_FILE_STORAGE_KEY, CONFIG_QUERY_KEY, PersonalShelf } from '#src/config';
+import { ConfigError, SettingsError } from '#src/utils/error';
+import { getModule } from '#src/modules/container';
+
+// Use local storage so the override persists until cleared
+const storage = window.localStorage;
+
+@injectable()
+export default class AppController {
+  private readonly configService: ConfigService;
+  private isLoaded: boolean = false;
+
+  constructor(configService: ConfigService) {
+    this.configService = configService;
+  }
+
+  getConfigSource(settings: Settings | undefined) {
+    if (!settings) {
+      return '';
+    }
+
+    const urlParams = new URLSearchParams(window.location.search);
+    const configKey = urlParams.get(CONFIG_QUERY_KEY);
+
+    // Skip all the fancy logic below if there aren't any other options besides the default anyhow
+    if (!settings.UNSAFE_allowAnyConfigSource && (settings.additionalAllowedConfigSources?.length || 0) <= 0) {
+      return settings.defaultConfigSource;
+    }
+
+    if (configKey !== null) {
+      // If the query param exists but the value is empty, clear the storage and allow fallback to the default config
+      if (!configKey) {
+        storage.removeItem(CONFIG_FILE_STORAGE_KEY);
+        return settings.defaultConfigSource;
+      }
+
+      // If it's valid, store it and return it
+      if (this.isValidConfigSource(configKey, settings)) {
+        storage.setItem(CONFIG_FILE_STORAGE_KEY, configKey);
+        return configKey;
+      }
+
+      logDev(`Invalid app-config query param: ${configKey}`);
+    }
+    // Yes this falls through from above to look up the stored value if the query string is invalid and that's OK
+
+    const storedSource = storage.getItem(CONFIG_FILE_STORAGE_KEY);
+
+    // Make sure the stored value is still valid before returning it
+    if (storedSource) {
+      if (this.isValidConfigSource(storedSource, settings)) {
+        return storedSource;
+      }
+
+      logDev('Invalid stored config: ' + storedSource);
+      storage.removeItem(CONFIG_FILE_STORAGE_KEY);
+    }
+
+    return settings.defaultConfigSource;
+  }
+
+  initAppSettings = async () => {
+    const settings = await fetch('/.webapp.ini')
+      .then((result) => result.text())
+      .then((iniString) => ini.parse(iniString) as Settings)
+      .catch((e) => {
+        logDev(e);
+        // It's possible to not use the ini settings files, so an error doesn't have to be fatal
+        return {} as Settings;
+      });
```

Comment:
I suggest we extract the setting's logic even further. Then it would be easier to replace only the settings part for a different platform without duplicating all full AppController.

Reply:
Agreed. I think the 'app' should just take a config ID input.

Reply:
Will do